File: services/nuldoc/nuldoc-src/pages/TagPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import BlogGlobalHeader from "../components/BlogGlobalHeader.ts";
import SlidesGlobalHeader from "../components/SlidesGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import PostPageEntry from "../components/PostPageEntry.ts";
import SlidePageEntry from "../components/SlidePageEntry.ts";
import { Config, getTagLabel } from "../config.ts";
import { getPostPublishedDate } from "../generators/post.ts";
import { TaggedPage } from "../generators/tagged_page.ts";
import { elem, Element, h1, header } from "../dom.ts";

export default async function TagPage(
  tagSlug: string,
  pages: TaggedPage[],
  site: "blog" | "slides",
  config: Config,
): Promise<Element> {
  const tagLabel = getTagLabel(config, tagSlug);
  const pageTitle = `タグ「${tagLabel}」一覧`;

  const GlobalHeader = site === "blog" ? BlogGlobalHeader : SlidesGlobalHeader;

  return await PageLayout({
    metaCopyrightYear: getPostPublishedDate(pages[pages.length - 1]).year,
    metaDescription: `タグ「${tagLabel}」のついた記事またはスライドの一覧`,
    metaKeywords: [tagLabel],
    metaTitle: `${pageTitle}｜${config.sites[site].siteName}`,
    metaAtomFeedHref: `https://${
      config.sites[site].fqdn
    }/tags/${tagSlug}/atom.xml`,
    site,
    config,
    children: elem(
      "body",
      { class: "list" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        header({ class: "page-header" }, h1({}, pageTitle)),
        ...pages.map((page) =>
          "event" in page
            ? SlidePageEntry({ slide: page, config })
            : PostPageEntry({ post: page, config })
        ),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/tag_page.rb
module Nuldoc
  module Pages
    class TagPage
      def initialize(tag_slug:, pages:, site:, config:)
        @tag_slug = tag_slug
        @pages = pages
        @site = site
        @config = config
      end

      def render
        config = @config
        site = @site
        pages = @pages
        tag_label = config.tag_label(@tag_slug)
        page_title = "タグ「#{tag_label}」一覧"

        global_header = site == 'blog' ? Components::BlogGlobalHeader : Components::SlidesGlobalHeader
        site_entry = config.site_entry(site)

        Components::PageLayout.new(
          meta_copyright_year: GeneratorUtils.published_date(pages.last).year,
          meta_description: "タグ「#{tag_label}」のついた記事またはスライドの一覧",
          meta_keywords: [tag_label],
          meta_title: "#{page_title}｜#{site_entry.site_name}",
          meta_atom_feed_href: "https://#{site_entry.fqdn}/tags/#{@tag_slug}/atom.xml",
          site: site,
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'list' do
              render global_header, config: config
              main class: 'main' do
                header(class: 'page-header') { h1 { text page_title } }
                pages.each do |page|
                  if page.respond_to?(:event)
                    render Components::SlidePageEntry, slide: page, config: config
                  else
                    render Components::PostPageEntry, post: page, config: config
                  end
                end
              end
              render Components::GlobalFooter, config: config
            end
          end
        ).build
      end
    end
  end
end
